### PR TITLE
dialects: (seq) Add ClockType and ClockDivider

### PR DESCRIPTION
**Pull request by lucjaulmes**
`seq` dialect first part, as discussed on zulip.

Opened as WIP but looking for feedback. Pinging @superlopuh @tobiasgrosser @compor and @JosseVanDelm as having participated in the discussion on zulip. Not sure who’s signing off on changes down the line.  
There seems to be quite a lot to discuss just on this first type/operation
- One thing nagging me in particular is that printing the `ClockDivider`’s power of 2 attribute with `print_attribute` means we get `seq.clock_div %clk by #int<4>` instead of the input syntax `seq.clock_div %clk by 4` -- should I print the naked `int` value to get the right syntax instead?  It seems to be parsed back fine.
- I haven’t managed to coerce this `pow2` attribute to be signless 8-bit, no matter how much I’ve been messing around with `IntAttr`, `IntegerType`, and other similarly named things. It always seems to be read by `parse_attribute()` as a 32 or 64 bit int -- suggestions welcome here.

Some questions that will need answering before any of this can go forward:

1. What’s the source of truth for a this dialect: CIRCT docs or code? They disagree in major ways, most notably docs define [`ClockType`](https://circt.llvm.org/docs/Dialects/Seq/#type-definition) with syntax `!seq.clock`, and use it throughout operations. This does not exist in the CIRCT codebase and clocks are implemented with `i1`.
2. A number of things in `seq` seem to depend on the `hw` dialect which is not in xdsl at this time. Most CIRCT tests of `seq` things use a `hw.module({})` region, and `hw.constant` values etc. More concretely, the attribute `::circt::hw::InnerSymAttr` is used in a number of  `seq` operations. There are a number of ways forward:
   a. Define a temporary `InnerSymAttr` in `seq` to be replaced by the proper import if/when `hw` is implemented,
   b. Implement operations using an `InnerSymAttr` without that attribute,
   c. Skip all operations using an `InnerSymAttr`,
   d. Create an empty `hw` dialect with just the needed attribute,
   e. Implement `hw` dialect,
   f. Something I didn’t think of?

### CODE REVIEW

**Comment by lucjaulmes** [comment]
Hmm looks like I missed the `pyright` set of tests. I’ll starting looking into that first.

**Comment by luisacicolini** [comment]
Based on the (little) experience I had with CIRCT i would suggest using the code as a main source!

File: tests/filecheck/dialects/seq/seq_invalid.mlir
Comment (superlopuh) [inline tests/filecheck/dialects/seq/seq_invalid.mlir:1]:
```suggestion
// RUN: xdsl-opt --verify-diagnostics --split-input-file %s | filecheck %s
```

File: tests/filecheck/dialects/seq/seq_invalid.mlir
Comment (superlopuh) [inline tests/filecheck/dialects/seq/seq_invalid.mlir:6]: I haven't seen this expected-error syntax, where is that from? Is it supported by our filecheck?

```suggestion
  // CHECK: {{Operation does not verify: divider value 5 is not a power of 2}}
```

File: xdsl/dialects/seq.py
Comment (superlopuh) [inline xdsl/dialects/seq.py:51]:
```suggestion
    pow2 = attr_def(AnyIntegerAttr)
```
You can just drop this

File: xdsl/dialects/seq.py
Comment (superlopuh) [inline xdsl/dialects/seq.py:55]: We require type annotations everywhere
```suggestion
    def __init__(self, clockIn: SSAValue | Operation, pow2: AnyIntegerAttr):
```

File: xdsl/dialects/seq.py
Comment (superlopuh) [inline xdsl/dialects/seq.py:77]: Yeah I think just printing the integer is good, not the attribute

**Review by superlopuh — CHANGES_REQUESTED** [review]
Looking great, would be good to start planning adding interop tests with circt so that we can be sure that they are actually interoperable, potentially worth creating a GitHub project to track what needs to be done for that to happen

File: tests/filecheck/dialects/seq/seq_invalid.mlir
Comment (lucjaulmes) [reply tests/filecheck/dialects/seq/seq_invalid.mlir:6]: It’s copied from CIRCT, might be wrong. `lit` didn’t complain but can update

**Comment by compor** [comment]
@lucjaulmes just a quick response before I look into this in a bit more detail;

Regarding your point 2., are we aiming for completeness, i.e., fully implementing  `seq` or can we get away with just enough operations?

Without this, I'd say implement `InnerSymAttr` within `seq` adding a TODO and a gh issue tracking this (the gh issue after this is merged).
For `hw.module` I'd ignore it and just use `builtin.module` for now.
For `hw.constant` I'd use `arith.constant` for now.

**Comment by tobiasgrosser** [comment]
One of the key conclusions with earlier PRs was that the best step is to look at a minimal set of operations to integrate, and to leave expanding to more operations for future PRs.

**Comment by superlopuh** [comment]
I'd maybe have the opposite suggestion to @compor and first add `hw.module` and `hw.constant` first because they seem to be trivial, and adding more interesting/complex operations as a follow-up, but maybe I'm missing something.

File: tests/filecheck/dialects/seq/seq_invalid.mlir
Comment (compor) [reply tests/filecheck/dialects/seq/seq_invalid.mlir:6]: hmm, I think `filecheck` has the `--check-prefix FOO` but it always appends `CHECK` to whatever you provide it, so I don't think the above prefix triggers anything.

File: xdsl/dialects/seq.py
Comment (compor) [inline xdsl/dialects/seq.py:61]: This is great.
You might also want to check that the `IntegerType` provided has the right width, etc. with something like:

```
if self.pow2.type != IntegerType(8):
  raise VerifyException("etc")
```

**Review by compor — CHANGES_REQUESTED** [review]
That's great, going over the `seq` documentation I think this is the smallest type/ops one could add in one go.

File: xdsl/dialects/seq.py
Comment (superlopuh) [inline xdsl/dialects/seq.py:61]:
```suggestion
        if self.pow2.value.data.bit_count() != 1:
```

**Comment by superlopuh** [comment]
This looks great! I think it's ready to review/merge, is there anything else that you're planning to add?

File: xdsl/dialects/seq.py
Comment (lucjaulmes) [reply xdsl/dialects/seq.py:30]: Ah, that’s a recent addition indeed. I only had `HLMemType` and `FirMemType` in that file. The [PR introducing it](https://www.github.com/llvm/circt/pull/5919) even says:
> Subsequent PRs will switch `seq` ops over to use [ClockType] and generate it from FIRRTL.

File: xdsl/dialects/seq.py
Comment (lucjaulmes) [reply xdsl/dialects/seq.py]: Didn’t know python had a proper popcount since 3.10 :roll_eyes:

File: tests/filecheck/dialects/seq/seq_ops.mlir
Comment (lucjaulmes) [reply tests/filecheck/dialects/seq/seq_ops.mlir:7]: Are you sure? Removing it makes the test fail, and it seems to be quite a common pattern:

```
> git grep -l XDSL_ROUNDTRIP  -- *.mlir | xargs grep -c '// CHECK:' | column -ts:
tests/filecheck/dialects/affine/examples.mlir                    2
tests/filecheck/dialects/arith/arith_attrs.mlir                  1
tests/filecheck/dialects/arith/arith_ops.mlir                    1
tests/filecheck/dialects/arith/arith_ops_custom.mlir             1
tests/filecheck/dialects/builtin/attrs.mlir                      16
tests/filecheck/dialects/builtin/module.mlir                     3
tests/filecheck/dialects/builtin/unrealized_conv_cast.mlir       8
tests/filecheck/dialects/cf/cf_ops.mlir                          4
tests/filecheck/dialects/cmath/cmath_ops.mlir                    2
tests/filecheck/dialects/comb/comb_ops.mlir                      1
tests/filecheck/dialects/dmp/ops.mlir                            1
tests/filecheck/dialects/fsm/fsm_op.mlir                         65
tests/filecheck/dialects/func/func_ops.mlir                      7
tests/filecheck/dialects/gpu/ops.mlir                            1
tests/filecheck/dialects/irdl/cmath.irdl.mlir                    4
tests/filecheck/dialects/irdl/cyclic-types.irdl.mlir             4
tests/filecheck/dialects/irdl/test-type.irdl.mlir                4
tests/filecheck/dialects/irdl/testd.irdl.mlir                    11
tests/filecheck/dialects/llvm/array.mlir                         1
tests/filecheck/dialects/llvm/example.mlir                       5
tests/filecheck/dialects/llvm/global.mlir                        1
tests/filecheck/dialects/llvm/pointers.mlir                      1
tests/filecheck/dialects/memref/memref_ops.mlir                  0
tests/filecheck/dialects/pdl/mlir-tests.mlir                     10
tests/filecheck/dialects/printf/printf_basics.mlir               1
tests/filecheck/dialects/riscv/riscv_ops.mlir                    2
tests/filecheck/dialects/riscv_scf/ops.mlir                      1
tests/filecheck/dialects/scf/reduce.mlir                         1
tests/filecheck/dialects/scf/scf_ops.mlir                        3
tests/filecheck/dialects/seq/seq_ops.mlir                        1
tests/filecheck/dialects/snitch/snitch_ops.mlir                  5
tests/filecheck/dialects/snitch_runtime/snitch_runtime_ops.mlir  40
tests/filecheck/dialects/stencil/stencil_ops.mlir                5
tests/filecheck/dialects/vector/vector_ops.mlir                  1
```

**Comment by compor** [comment]
LGTM.
I'd move it as ready to review (remove WIP title tag and the draft status).
If there's nothing else to add, ping us to approve and once a couple of ticks have been given, feel free to merge.

**Review by compor — APPROVED** [review]

File: tests/filecheck/dialects/seq/seq_ops.mlir
Comment (compor) [inline tests/filecheck/dialects/seq/seq_ops.mlir:3]:
```suggestion
```

File: xdsl/dialects/seq.py
Comment (compor) [inline xdsl/dialects/seq.py:59]: Can this condition be tested too?

**Review by compor — APPROVED** [review]

**Review by superlopuh — APPROVED** [review]

File: tests/dialects/test_seq.py
Comment (compor) [inline tests/dialects/test_seq.py:22]: This is great.
I'd simplify by not using the builder methods and using the `test` dialect for whatever is not strictly needed for testing.

```suggestion
    clock_div = ClockDivider(
        test.TestOp.create(result_types=[clock]), IntegerAttr(513, i32)
    )
```

**Review by compor — APPROVED** [review]

File: tests/dialects/test_seq.py
Comment (superlopuh) [reply tests/dialects/test_seq.py:22]: nit:
```suggestion
    clock_div = ClockDivider(
        test.TestSSAValue(clock), IntegerAttr(513, i32)
    )
```

**Comment by compor** [comment]
@lucjaulmes go ahead and merge?

**Comment by lucjaulmes** [comment]
Go for it @compor. I don’t have write access.

**Pull request merged**
